File: 0139-word-break/solution.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        dp = [False] * (len(s) + 1)
        dp[len(s)] = True
        for index in range(len(s) - 1, -1, -1):
            for curWord in wordDict:
                if len(s) >= (len(curWord) + index) and s[index:index+len(curWord)] == curWord:
                    dp[index] = dp[len(curWord) + index]
                if dp[index]:
                    break
        return dp[0]

File: 0139-word-break/solution.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        int n = s.size();
        vector<int> dp(n + 1, false);
        dp[n] = true;
        for(int i = n - 1; i >= 0; --i){
            for(const auto& word : wordDict){
                if(i + word.size() <= n and dp[i + word.size()] == true and s.substr(i, word.size()) == word){
                    dp[i] = true;
                }
            }
        }
        return dp[0];
    }
};
